Dockerfile

```diff
@@ -2,99 +2,97 @@
+RUN if [ "$PULLREQ" = "false" ] && [ "$BRANCH" = "master" ]; then \
       echo "No pull request number given - defaulting to master branch"; \
     elif [ "$BRANCH" != "master" ]; then \
       cd navigation2; \
       git fetch origin $BRANCH:temp_branch; \
       git checkout temp_branch; \
-      cd -; \
       echo "No pull request number given - defaulting to $BRANCH branch"; \
     else \
+      cd navigation2; \
       git fetch origin pull/$PULLREQ/head:pr_branch; \
       git checkout pr_branch; \
     fi
 
-# Download dependencies
-RUN echo "Downloading the ROS 2 navstack dependencies workspace"
-WORKDIR /ros2_ws/navstack_dependencies_ws/src
-RUN vcs import . < /ros2_ws/navigation2_ws/src/navigation2/tools/ros2_dependencies.repos
+# clone dependency package repos
+ENV ROS_WS /opt/ros_ws
+RUN mkdir -p $ROS_WS/src
+WORKDIR $ROS_WS
+RUN vcs import src < $NAV2_WS/src/navigation2/tools/ros2_dependencies.repos
+
+# install dependency package dependencies
+RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
+    apt-get update && \
+    rosdep install -q -y \
+      --from-paths \
+        src \
+      --ignore-src \
+    && rm -rf /var/lib/apt/lists/*
+
+# build dependency package source
+ARG CMAKE_BUILD_TYPE=Release
+RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
+    colcon build \
+      --symlink-install \
+      --cmake-args \
+        -DCMAKE_BUILD_TYPE=$CMAKE_BUILD_TYPE
 
-# Build ROS 2 dependencies
-WORKDIR /ros2_ws/navstack_dependencies_ws
-RUN rosdep init && rosdep update && rosdep install -q -y -r --from-paths src --ignore-src --rosdistro $ROS2_DISTRO --as-root=apt:false --as-root=pip:false --skip-keys "catkin"
-# source
-RUN (source /opt/ros/$ROS2_DISTRO/setup.bash && . /ros2_ws/ros2-linux/setup.bash && colcon build --symlink-install)
+# install navigation2 package dependencies
+WORKDIR $NAV2_WS
+RUN . $ROS_WS/install/setup.sh && \
+    apt-get update && \
+    rosdep install -q -y \
+      --from-paths \
+        $ROS_WS/src \
+        src \
+      --ignore-src \
+    && rm -rf /var/lib/apt/lists/*
 
-# Build navigation2 code
-WORKDIR /ros2_ws/navigation2_ws
-RUN rm /ros2_ws/navigation2_ws/src/navigation2/nav2_system_tests/COLCON_IGNORE
-RUN rosdep install -q -y -r --from-paths src --ignore-src --rosdistro $ROS2_DISTRO --as-root=apt:false --as-root=pip:false
-RUN (. /ros2_ws/navstack_dependencies_ws/install/setup.bash && colcon build --symlink-install)
+# build navigation2 package source
+RUN rm $NAV2_WS/src/navigation2/nav2_system_tests/COLCON_IGNORE
+RUN . $ROS_WS/install/setup.sh && \
+    colcon build \
+      --symlink-install \
+      --cmake-args \
+        -DCMAKE_BUILD_TYPE=$CMAKE_BUILD_TYPE
 
-RUN echo "export ROS_DOMAIN_ID=22" >> /root/.bashrc
-COPY tools/ctest_retry.bash /ros2_ws/navigation2_ws/build/nav2_system_tests
+# source navigation2 workspace from entrypoint
+RUN sed --in-place \
+      's|^source .*|source "$NAV2_WS/install/setup.bash"|' \
+      /ros_entrypoint.sh
 
-CMD ["bash"]
+ENV ROS_DOMAIN_ID 22
```

Comment:
What does this buy us?

Reply:
I couldn't get the system test to pass without this, suggesting that somehow different systems were interfering with each other. WIthout this variable set, all ROS 2 systems on the network can see each other. This variable makes it so only the ROS 2 systems with ID 22 can see each other. 

I'm assuming that somewhere at Travis, at any given moment there must be some ROS 2 node running causing interference. 

Anyhow, it's an ugly hack. Clearly the system test needs work. Maybe I'll just disable it again for the time being.

Reply:
Is there a reason you need to enable the `host` network driver? I.e: `docker run ... --net=host`. I feel this is unnecessary when only using one container.

Reply:
Well ... I tried it with and without setting that flag, and it only worked with it set. My assumption was that the network restrictions in docker were interfering with the ROS 2 infrastructure, but I never investigated why exactly.

As I recall, AMCL couldn't receive a map from the map server without that set. This is the first communication between elements of the navigation stack, so it seemed even the most basic message passing was inoperable without `--net=host`

However, this could stand more investigation.

Reply:
Now that the system test is running again, I retried without --net=host and without setting the ROS_DOMAIN_ID and the system test seems to work fine.

I'm not sure why they seemed necessary before. All I can think of is that, since the test was flaky, I just got unlucky and got random passes with those set and got random failures when they weren't.